aclk/aclk_query_queue.c

```diff
@@ -45,14 +45,49 @@ static inline int _aclk_queue_query(aclk_query_t query)
 
 }
 
+// Gets a pointer to the metric associated with a particular query type.
+// NULL if the query type has no associated metric.
+static inline volatile uint32_t *aclk_stats_qmetric_for_qtype(aclk_query_type_t qtype) {
+    switch (qtype) {
+        case HTTP_API_V2:
+            return &aclk_metrics_per_sample.cloud_req_type_http;
+        case ALARM_STATE_UPDATE:
+            return &aclk_metrics_per_sample.cloud_req_type_alarm_upd;
+        case METADATA_INFO:
+            return &aclk_metrics_per_sample.cloud_req_type_metadata_info;
+        case METADATA_ALARMS:
+            return &aclk_metrics_per_sample.cloud_req_type_metadata_alarms;
+        case CHART_NEW:
+            return &aclk_metrics_per_sample.cloud_req_type_chart_new;
+        case CHART_DEL:
+            return &aclk_metrics_per_sample.cloud_req_type_chart_del;
+        case REGISTER_NODE:
+            return &aclk_metrics_per_sample.cloud_req_type_register_node;
+        case NODE_STATE_UPDATE:
+            return &aclk_metrics_per_sample.cloud_req_type_node_upd;
+        default:
+            return NULL;
+    }
+}
+
 int aclk_queue_query(aclk_query_t query)
 {
     int ret = _aclk_queue_query(query);
     if (!ret) {
+        // local cache of query type before we wake up query thread, which may
+        // free the query in a race.
+        aclk_query_type_t qtype = query->type;
         QUERY_THREAD_WAKEUP;
+
         if (aclk_stats_enabled) {
+            // get target query type metric before lock so we keep lock for
+            // minimal time.
+            volatile uint32_t *metric = aclk_stats_qmetric_for_qtype(qtype);
```

Comment:
I have so many questions about the use of `volatile`. Why do we use it throughout the code? Why and/or where do we not update stats in locked-regions? Did we profile the code to check that `aclk_stats_qmetric_for_qtype`'s latency (which should be in the order of nanoseconds) justifies moving this outside of the lock?

I won't block the review because of this, but I think we really should post a follow up patch to use either locks or (builtin even) atomics when updating stats.

Reply:
Those are reasonable questions @vkalintiris.

I'd imagine `volatile` is being used to avoid dealing with stale reads in a multi-threaded setting when the compiler tries to pull tricks. But you're right, we take locks, obviously for writes, and in fact even for reads when reading to insert into the timeseries db. In that case we cannot possibly be dealing with subtle data races across threads, so in the future usage of `volatile` here could be removed.

I did not profile the code to check the latency of `aclk_stats_qmetric_for_qtype`, which you're right would be often just a few cmp instructions and unnoticeable. But then again, there is no justification to move it back into the lock either - it's a free win with no effort.

I agree we should probably use atomics to update stats. I don't think the cost of a pthread mutex, which could sleep, is reasonable to pay for doing arithmetic instructions safely.

Reply:
Thanks @UmanShahzad!

AFAIK, ops on volatiles are not atomic and their ordering is defined relative to each other only within the context of a single thread (unless they are guarded by a lock). I asked those questions mostly to find out if we don't really care about displaying 100% accurate information 100% of the time. If this is the case:
- I guess we can get away with it for the time being, but we still should fix this at some point, either by moving these ops in locked-region or by using atomic ops.
- I'm curious if there is a quick/easy way to verify that the gcc/clang thread sanitizer does not complain. If it does, we can either prioritize a little bit more the aforementioned changes, or add some kind of suppression file.

Reply:
Oh, I think there may be a misunderstanding here @vkalintiris : all changes to the stats are already in locked regions. This particular line we're commenting on is just getting the address of a particular stat variable, but updating it only later in the lock. The point is that which stat to update depends upon the query type, so we pick the right one by address outside of the lock region, then update it while locked.

So we're definitely getting 100% accurate info 100% of the time right now, whether here or elsewhere. But long-term, given that a pthread lock is almost always >= cost of a single atomic update (since it too is just an atomic CAS), we should probably move to atomics for updating/reading simple stats.

Reply:
it is not so already to not have both implemented `#if defined(HAVE_C___ATOMIC) && !defined(NETDATA_NO_ATOMIC_INSTRUCTIONS)` but yeah atomics are much better and faster way to deal with this if available

Reply:
Thanks for the feedback, folks! I have to admit I'm a little bit confused. If read/writes are synchronized, why do we need `volatile` at all? Can't we just get a non-volatile object's address outside of locks?

Reply:
@vkalintiris For this particular line, we need `volatile`, because the data being pointed to is `volatile`.

Longer term we need to change the underlying stat data to not be volatile. We can probably do that right now actually but definitely out of this PR!

(We might also be missing something because I see other parts of the codebase do similar things with `volatile`. I don't know, this is something to ask someone who knows the reasoning for the original cause of this; `git blame` shows that it might have been Costa).